lib/chain.js

```diff
@@ -2,82 +2,46 @@
 
 module.exports = (api) => {
 
-  const resolver = (resolve, reject) => (
-    (err, result) => {
-      if (err) reject(err);
-      else resolve(result);
-    }
-  );
-
-  class ArrayChain {
-
-    constructor(array) {
-      this._promise = Promise.resolve(array);
-    }
-
-    then(fn) {
-      return this._promise.then(fn);
-    }
-
-    catch(fn) {
-      return this._promise.catch(fn);
-    }
-
-    map(fn) {
-      this._update((array, resolve, reject) => {
-        api.metasync.map(array, fn, resolver(resolve, reject));
-      });
-      return this;
-    }
-
-    filter(fn) {
-      this._update((array, resolve, reject) => {
-        api.metasync.filter(array, fn, resolver(resolve, reject));
-      });
-      return this;
-    }
-
-    reduce(fn, initial) {
-      this._update((array, resolve, reject) => {
-        api.metasync.reduce(array, fn, resolver(resolve, reject), initial);
-      });
-      return this._promise;
-    }
-
-    each(fn) {
-      this._update((array, resolve, reject) => {
-        api.metasync.each(array, fn, resolver(resolve, reject));
-      });
-      return this._promise;
-    }
-
-    series(fn) {
-      this._update((array, resolve, reject) => {
-        api.metasync.series(array, fn, resolver(resolve, reject));
-      });
-      return this._promise;
-    }
-
-    find(fn) {
-      this._update((array, resolve, reject) => {
-        api.metasync.find(array, fn, resolver(resolve, reject));
-      });
-      return this._promise;
-    }
-
-    _update(fn) {
-      this._promise = this._promise.then((array) => (
-        new Promise((resolve, reject) => fn(array, resolve, reject))
-      ));
-    }
-
+  function arrayChain(array, prev = null) {
+
+    let next = null, done = null, fail = null;
+
+    const self = (err, data) => {
+      array = data;
+      if (next) next();
+      if (err) {
+        if (fail) fail(err);
+      } else if (done) {
+        done(data);
+      }
+    };
+
+    if (!prev) process.nextTick(() => self(null, array));
+
+    self.then = (fn) => (done = fn, self);
+    self.catch = (fn) => (fail = fn, self);
+
+    const chain = (performer) => (fn, initial) => {
+      const res = arrayChain(null, self);
+      next = () => performer(array, fn, res, initial);
+      return res;
+    };
+
+    self.map = chain(api.metasync.map);
```

Comment:
When I posted that https://github.com/metarhia/MetaSync/pull/61#discussion_r109326647 I was thinking more about
```javascript
const methods = ['map', 'filter', 'reduce', 'each', 'series', 'find'];

for (let i = 0; i < methods.length; i++) {
  const method = methods[i];
  self[method] = chain(api.metasync[method]);
}
```
I am not using for-in loop because you were concerned about performance in this case and I don't exactly know whether it will hurt performance or not.